pyparsing/core.py

```diff
@@ -1504,64 +1504,43 @@ def __mul__(self, other) -> "ParserElement":
+        elif isinstance(other, tuple):
+            minElements, maxElements = (other + (None, None))[:2]
+        elif other is Ellipsis:
+            minElements, maxElements = other, None
         else:
-            other = tuple(o if o is not Ellipsis else None for o in other)
-            other = (other + (None, None))[:2]
-            if other[0] is None:
-                other = (0, other[1])
-            if isinstance(other[0], int) and other[1] is None:
-                if other[0] == 0:
-                    return ZeroOrMore(self)
-                if other[0] == 1:
-                    return OneOrMore(self)
-                else:
-                    return self * other[0] + ZeroOrMore(self)
-            elif isinstance(other[0], int) and isinstance(other[1], int):
-                minElements, optElements = other
-                optElements -= minElements
-            else:
-                return NotImplemented
+            return NotImplemented
 
-        if minElements < 0:
+        if minElements in (Ellipsis, None):
+            minElements = 0
+        elif type(minElements) != int:
+            return NotImplemented
+        elif minElements < 0:
             raise ValueError("cannot multiply ParserElement by negative value")
-        if optElements < 0:
-            raise ValueError(
-                "second tuple value must be greater or equal to first tuple value"
-            )
-        if minElements == optElements == 0:
-            return And([])
-
-        if optElements:
 
-            def makeOptionalList(n):
-                if n > 1:
-                    return Opt(self + makeOptionalList(n - 1))
-                else:
-                    return Opt(self)
-
-            if minElements:
-                if minElements == 1:
-                    ret = self + makeOptionalList(optElements)
-                else:
-                    ret = And([self] * minElements) + makeOptionalList(optElements)
+        if maxElements in (Ellipsis, None):
+            if minElements == 0:
+                return ZeroOrMore(self)
+            elif minElements == 1:
+                return OneOrMore(self)
             else:
-                ret = makeOptionalList(optElements)
-        else:
-            if minElements == 1:
-                ret = self
+                return self * minElements + ZeroOrMore(self)
+        elif type(maxElements) != int:
+            return NotImplemented
+        elif maxElements < minElements:
+            raise ValueError("second tuple value must be greater or equal to first tuple value")
+
+        if minElements == maxElements:
+            if minElements == 0:
+                return And([])
+            elif minElements == 1:
+                return self
             else:
-                ret = And([self] * minElements)
-        return ret
+                return And([self] * minElements)
+
+        return And([self] * minElements) + (Opt(self) * (maxElements - minElements))
```

Comment:
No hurry on code review; but from re-inspecting this myself, I'm beginning to wonder about this `Opt` construction and the effect this could have on the parse output.

In particular: perhaps it's better to have a chain of nested `Opt` nodes (recursive, instead of iterative as in this PR), because then matching an input against the parse tree can exit as soon as a single token _doesn't_ match -- whereas the `And[Opt(...), Opt(...), ...]` formulation here would be evaluated in full (I think?).